04-cs-fundamentals/internship/exercises/binary-tree.md

```diff
@@ -0,0 +1,67 @@
+# Binary Tree Exercises
+
+## Instructions
+
+In pairs answer the following questions as if they are whiteboarding exercises.  Write down your answers for each step.
+
+For Each Question:
+
+- **Step 1:**  Write down 2 - 3 clarifying questions you would ask an interviewer.
+- **Step 2:**  Write down assumptions for each of those questions.
+- **Step 3:**  Generate 3+ sample input & output combinations to the methods described.  Look for edge-cases
+- **Step 4:**  Write a proposed solution.
+- **Step 5:**  Test your solution with your sample input.
+- **Step 6:**  Evaluate your solution in terms of space & time complexity.
+
+## Sample Code
+
+```ruby
+class Student
+  attr_reader :id, :name
+
+  def initialize(id, name)
+    @id = id
+    @name = name
+  end
+end
+
+class TreeNode
+  attr_reader :key, :value
+  attr_accessor :left, :right
+
+  def initialize(k, v, l = nil, r = nil)
```

Comment:
Having full argument names would be nice.

Assuming you want to be nice, that is.